### ISSUE TITLE
cluster_api - select using one more more indices

### ISSUE THREAD

**Issue by tsafin** [tarantool/vshard-cluster-api#3]
* `select` is based upon function execution on all or part of nodes in the cluster, i.e. essentially it's MapReduce operation. 
* there is no need to pass all returned data from some node, it's usually excessive, thus we need to provide _pagination_ mechanism for partial data return. i.e. approximately interface might be following:
 
```
local iter, err = map_reduce(func_name, args, opts)
```
 
Fucntion returns iterator, whcih hides complex logics behing facade. Iterator inetrface might be looking like:
 
- `has_next()`
- `get()`
- `next()`
- `dispose()`
 
e.g.:
 
```
while iter.has_next() do
     local item, err = iter.get()
     if err ~= nil …
     iter.next()
end
```

Map-reduce operation might be terminated at any moment via `iter.dispose()`, after then `has_next` would always return `false`

**Comment by tsafin** [comment]
Дальше я сломался переводить, и просто процитирую Макса @mtrempoltsev 

**Что происходит за фасадом?**
 
При вызове map_reduce создается уникальный id и вызывается из различных файберов map_reduce_begin на каждой ноде кластера.
 
Далее на каждой ноде вызывается указанная в вызове map_reduce функция и в таблице по id создается state. Перед тем как обсудить, что это за state хочу ввести понятие resumable функции - это функция, которая при вызове принимает объект со state и модифицирует его. Примером такой функции может быть функция которая, если state не установлен, то берет итератор по индексу, выбирает 10 таплов, возвращает их и кладет итератор в state. При следующем вызове она выбирает следующие 10 элементов и так далее. resumable функция или нет указывается, например, опцией при вызове map_reduce.
 
Таким образом в паре id и state, state является тем, что передается resumable функции при каждом вызове, либо это результат вызова не resumable функции.
 
Зачем это понадобилось? Нужно это для того чтобы ограничить количество передаваемых по сети данных, так на каждой ноде пользовательская функция может вернуть, скажем, 100 таплов, у нас 20 нод, итого 2000 таплов, при этом вызывающему коду достаточно первых 10. Поэтому мы передаем с каждой ноды по 10 таплов и сохраняем state, если этого будет недостаточно отдаем при следующем запросе следующие 10 (значение взято с потолка).
 
На стороне вызова мы получаем Н ответов по числу нод и далее простым алгоритмом через посредник итератор их отдаем:
 
1. Положили в кучу с каждой ноды пару из тапла и номера ноды
2. На каждой итерации берем из кучи верхушку и возвращаем, в кучу кладем новый тапл от ноды по номеру, если таплов нет, то фоном делаем запрос на ноду map_reduce_continue с id сгенерированным при вызове map_reduce, результат используем при следующих итерациях
 
При вызове iter.dispose() или когда на каждой из нод данных больше нет, вызываем на всех нодах map_reduce_finish чтобы очистить память. Также стоит чистить на нодах память если был разорван конекшн с вызывающей нодой.
 
В плане эффективности особое внимание следует уделить на исключении преобразований из мсгпака в луа объекты и наоборот. Возможно на первом этапе это не удастся из-за интерфейса vshard, но в будущем это может быть исправлено.
 
Имея этот механизм становится возможен следующий шаг - реализация высокоуровневых функций доступа к данным. Например, select:
 
select("accounts", {{'=', 'acc_type', 'saving'}, {'>', 'amount', 0}}, { limit = 2 })
 
Здесь мы делаем выборку в кластере из спейсов accounts 2 тапла, где acc_type == saving и amount > 0. Для этого надо написать resumable функцию, которая обернет два итератора в один и позволит итерироваться по нему. Основная логика - мапинг условий в правильные итераторы.
 
На вызывающей стороне это обертка над map_reduce и итератором.
 
Сразу скажу об ограничениях:
 
1. Все условия только И, так как ИЛИ будет приводить к фулскану, поэтому лучше чтобы люди видя ограничения API думали о решении своей задачи в правильном ключе
2. Здесь может дать хороший профит реализация на С, у нас накоплен большой опыт в части оптимизации мап-редьюса в ТДГ - приходите за подробностями
3. В условиях используются только поля имеющие индекс - это просто для реализации, хотя в будущем наверняка захочется убрать это ограничение, но это потребует написание планировщика, а за ним и анализатор запросов. К слову, в ТДГ можно в поиске учитывать не только индекс
4. На первом этапе речь о полноценном планировщике не идет, хотя в будущем, конечно планировщик сможет дать существенную выгоду
 
API для вызова хранимок по-сути будет представлять собой тоже слегка обернутый уже описанный map_reduce.
 
По моему опыту, описанный фронт работ может занять пару разработчиков на 2-3 месяца для написания покрытой тестами и более-менее оптимизированной версии, поправьте меня если я ошибаюсь. Если у нас есть консенсус в понимании задачи, можно детализировать это в виде RFC и тасок.

**Comment by olegrok** [comment]
Я ещё раз перечитал написанное? И у меня возникает вопрос - уверены ли мы что нам нужно вводить какой-то стейт. Который ещё и вручную надо очищать?

Я всё понимаю, но LuaJIT однажды упрется в свои 2 гигабайта и на этом закончится весь map-reduce.

**Comment by tsafin** [comment]
> Я ещё раз перечитал написанное. И у меня возникает вопрос - уверены ли мы что нам нужно вводить какой-то стейт. Который ещё и вручную надо очищать?

Да, кажется, сейчас мы скорее уверены, что стейт не нужен и мы можем реализовать пагинацию без хранения состояния, при помощи ffi трюков с доступом к внутренностям итераторов.
